ES6/index.html

```diff
@@ -1,115 +1,47 @@
-                });
-            }
+    <div class="pager">${pager}</div>
+    <div class="paged-content">Page 1</div>
+    <div class="pager">${pager}</div>
+`;
         }
-        
-        class Pager 
-        {
-            constructor(elements, content)
-            {
-                this.elements = elements;
-                this.content = content;
-                
-                this.render();
-            }
-            
-            render()
-            {
-                let that = this;
-                let page = this.content.currentPage;
-                let pages = this.content.numberOfPages;
-                
-                function createPage(page, selected)
-                {
-                    let element = document.createElement("span");
-                    
-                    element.classList.add("page");
-                    
-                    if(selected) element.classList.add("selected");
-                    
-                    element.textContent = String(page);
-                    element.addEventListener("click", function() {
-                        that.content.currentPage = page;
-                        
-                        that.render();
-                    });
-                    
-                    return element;
-                }
-                
-                this.elements.forEach(function(element, index) {
-                    let title = document.createElement("span");
-                    
-                    title.textContent = "Page: ";
-                    
-                    element.replaceChildren(title);
-                    
-                    for(let i=0; i<pages; i++)
-                    {
-                        element.appendChild(createPage(i+1, (page === i+1)));
-                    }
-                });
-            }
+        connectedCallback() {
+          this.shadowRoot.querySelectorAll(".page").forEach((el) => {
+            if (el.dataset.page === "1") el.classList.add("selected");
+            el.addEventListener("click", this.navigate);
+          });
         }
-        
-        document.addEventListener("DOMContentLoaded", function() {
-            let content = new PagedContent(document.querySelectorAll("#content"));
-            let pager = new Pager(document.querySelectorAll(".pager"), content);
-        });
+
+        navigate = (ev) => {
+          this.shadowRoot.querySelector(".paged-content").textContent = `Page ${e.dataset.page}`;
+          this.shadowRoot.querySelectorAll(".page").forEach((el) =>
+              el.classList[el.dataset.page === e.dataset.page ? "add" : "remove"]("selected")
+            );
+        };
+      }
+      customElements.define("c-paged", Paged);
     </script>
-</head>
-<body>
-    <div id="top_pager" class="pager"></div>
-    <div id="content" class="paged-content"></div>
-    <div id="bottom_pager" class="pager"></div>
-</body>
+  </head>
+  <body>
+    <c-paged></c-paged>
```

Comment:
The content tag(s) should be separated from the pager tag(s) so that they can be mixed and matched. e.g. Some applications might do this:

```
<c-pager for="content"></c-pager>
<c-content id="content" pages="5"></c-content>
<c-pager for="content"></c-pager>
```

...and others might do just this:

```
<c-content id="content" pages="5"></c-content>
<c-pager for="content"></c-pager>
```

Reply:
this is an arbitrary requirement you've added post-hoc. This was never framed as component library snippet, if users want to arrange it differently, they can edit the component directly 🤨

Reply:
Look up the jQuery component model. The original code represents that model. The ES6 code uses Classes for the same purpose. 

You can change the original HTML exactly as I have given here. This will work fine:

```
<div id=“content”></div>
<div class=“pager”></div>
```

Reply:
You can also make that change in my code and it'll work fine
```js
shadow.innerHtml = `
   <!- snip -!>
    <div class="paged-content">Page 1</div>
    <div class="pager">${pager}</div>
`
```

the better api would be to use [slots](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_templates_and_slots) instead of arbitrary class names and ids if you're seeking a meaningfully reusable component out of this.

The jQuery example is quite bad from a component design imho, it basically _can't_ do components, only widgets. The difference between them as I use them here is composability. Of state, of logic, of children, of siblings, of templates. **Inherently** Components have all, widgets have none. You can also claim that components are declarative and widgets are imperative.
Your jQuery widget is not a component, it's unusable in a meaningful sense if you don't have control of its source directly in the editor. First of all because rendering the contents in the way you do you obliterate any contents the elements you hooked in had, and you destroy the entire view and reconstruct it on every interaction, detaching any other logic that was added. Ofc without any method to even hook into that lifecycle.

This is another reason why I used a shadow dom and a web component, since that mimics the render pattern you use without the pitfalls. You can't touch the controlled elements, and any accidental children still exist in the dom, but are not rendered (<- preserves state and so-on, failing elegantly instead of explosively)

I assume that you're very unfamiliar with web technologies, which is fine, I really don't mean that in an offensive way. Just like you'd be able to tell when some cpp abstraction is useful or helpful or required or insufficient, I am able to tell when web code is useful or helpful or required or insufficient. You probably didn't even consider your rendering approach beyond a rough estimation of a MVC/MVVM-type thing, which again, is **fine**, it's not your job. But these things matter, especially once you start to assert certain qualities your solution possesses which it doesn't